cmake/modules/RootMacros.cmake

```diff
@@ -609,7 +609,11 @@ function(ROOT_GENERATE_DICTIONARY dictionary)
 
     # get target properties added after call to ROOT_GENERATE_DICTIONARY()
     if(TARGET ${ARG_MODULE})
+if(MSVC)
```

Comment:
We should not special case MSVC here. @linev, is there a good way to reduce the duplicate `-I`? I remember you had something similar in a pull request but I not recall if it landed.

Reply:
> We should not special case MSVC here. @linev, is there a good way to reduce the duplicate `-I`? I remember you had something similar in a pull request but I not recall if it landed.

I agree, but the `if(MSVC)` is there only not to force `cmake_minimum_required(VERSION 3.15)` on all platforms

Reply:
I think we should run the deduplication step on the `INCLUDE_DIRECTORIES` property in cmake. I remember that Sergey was doing something about it.

Reply:
Any solution would be fine with me. We'll see once Sergey is back from vacation

Reply:
If such long arguments list is critical for the Windows - one have to apply patch, I see no other solution.
I also tried to reduce duplication, but does not succeed - because of generator expressions.
May be one can slightly modify patch and use `REMOVE_DUPLICATES` always when cmake version is supporting it.
But keep condition that Windows always require cmake version 3.16

Reply:
Can we not use `list(REMOVE_DUPLICATES)` over `INCLUDE_DIRECTORIES`?

Reply:
> Can we not use list(REMOVE_DUPLICATES) over INCLUDE_DIRECTORIES?

No. Include directories in some cases not yet configured when dictionary generation command is produced by cmake.
In many places in ROOT we have `target_include_directories(libname incdir1 incdir2)` after calling `ROOT_STANDARD_LIBRARY_PACKAGE`, where dictionary command is produced. Therefore we are using "generator expression" to extract these includes at the moment when cling is invoked. And therefore we need "generator expression" to remove duplicates at this moment. Thanks to @bellenot we have solution, but it requires very recent cmake.

Some time ago we have discussion with @Axel-Naumann to provide list of required includes directories into `ROOT_STANDARD_LIBRARY_PACKAGE` macro. In such case we can avoid clumsy generator expressions, but you and Axel refuse such changes. Therefore we have to use solution from Bertrand now.

Reply:
I think the problematic line is [here](https://github.com/root-project/root/blob/master/cmake/modules/RootMacros.cmake#L631). We seem to have the same content in `includedirs` (which is `incdirs` without duplicates) and `$<$<BOOL:${module_incs}>:-I$<JOIN:${module_incs},;-I>>`.

@Bertrand, I'd propose three options:
a) we delete the `includedirs` -- that might be breaking things outside ROOT.
b) deduplicate `includedirs` based on `module_incs` (the same holds for `definitions` vs `module_defs`) -- relatively cheap to implement.
c) option b) + a dev warning that we have already that value set in one of the variables -- that'd require reworking the setting of `incdirs` where @linev has most knowledge nowadays...

Reply:
> I think the problematic line is [here](https://github.com/root-project/root/blob/master/cmake/modules/RootMacros.cmake#L631). We seem to have the same content in `includedirs` (which is `incdirs` without duplicates) and `$<$<BOOL:${module_incs}>:-I$<JOIN:${module_incs},;-I>>`.
> 
> @Bertrand, I'd propose three options:
> a) we delete the `includedirs` -- that might be breaking things outside ROOT.
> b) deduplicate `includedirs` based on `module_incs` (the same holds for `definitions` vs `module_defs`) -- relatively cheap to implement.
> c) option b) + a dev warning that we have already that value set in one of the variables -- that'd require reworking the setting of `incdirs` where @linev has most knowledge nowadays...

Turns out that b) is not that easy because the `incdirs` contain some hard coded [here](https://github.com/root-project/root/blob/master/cmake/modules/RootMacros.cmake#L460) without being added as target_include_dirs. @linev any idea what was the motivation?

Reply:
> any idea what was the motivation

It is required by rootcling - main include directory must be on the first place.
As I remember - because of `runtime_cxxmodules`. 
If not specified, rootcling complains that same header can be located twice - in its original dirctory and in `$ROOTSYS/include`.

Reply:
I meant to ask why we just add the include folder to a list rather than calling target_include_directory?

Reply:
It is done to guarantee that `$ROOTSYS/include` is on the first place for dictionary
For library itself it is not required and was intentionally removed

Reply:
@linev, thanks!

@bellenot, go ahead with your current approach. What I am proposing will take a lot of time to implement as it is not so simple to deduplicate the information from both variables without significant changes to `GENERATE_ROOT_DICTIONARY`.

Reply:
@vgvassilev OK, thanks